### ISSUE TITLE
When will resignation be enabled?

### ISSUE THREAD

**Issue by isty2e** [gcp/leela-zero#135]
Apparently current LZ doesn't have any good idea about counting, and that has been the reason of disabling resignation for self-play games. A few questions regarding this:

1. When will be the point of allowing it to resign? LZ would probably remain weak for a long time, which also suggests that its winrate estimation will remain not that trustworthy. Even if it learns how to catch a stone, its prediction on winrates will not improve that much, I believe.
2. Will the AGZ-like approach, namely disallowing resignation for 10% of the games, be suitable at the early stage?
3. By allowing resignation, the individual games will be shorter in general, and there would be less board-filling games. Will it somehow affect the trained network?

**Comment by gcp** [comment]
AGZ pulled statistics from the games first to set a suitable resignation threshold. It's possible to obtain the same during validation tests, and hence see if it makes sense to try resigning, but this needs some tooling to be written first.

**Comment by isty2e** [comment]
If it makes sense, we probably should allow resignation as soon as possible. LZ is in its early stage, and refreshing games faster with updated networks would be a virtue. And it will presumably be trained less with nearly fully filled board, which might be desirable too.

**Comment by gcp** [comment]
This assumes the bot is estimating reliable who is winning. That's far from obvious. And if it's not, it needs to see the game played out and realize that it's wrong.

**Comment by isty2e** [comment]
Yeah, so that's why setting a proper resignation threshold depending on the network is important, if I understand correctly. Given the data of self-play games, how long will it take to estimate the resignation threshold for a trained NN?

**Comment by gcp** [comment]
You can't, see above. It needs tooling during the validation test.

**Comment by isty2e** [comment]
Oh well... I got it.

**Comment by Ttl** [comment]
I wrote some Python code to analyze the resign rate from the completed self-play games played with the newest network. I loaded completed games to leelaz, ran it with one playout and recorded the NN eval value for each position.

![min_winner_nn_eval](https://user-images.githubusercontent.com/544240/33237777-b9340cba-d285-11e7-8ad8-952c8d0c270e.png)

The above plot is the minimum NN eval value for the winner during the game. If the resign rate is set above the winner's minimum rate then there won't be an incorrect resignation.

Some statistics I calculated:

```
Dataset size 145 games
Minimum observed evaluation for winner 0.00377
Minimum evaluation in game for winner mean=0.20014442069, std=0.0866228330411
Minimum evaluation in game for loser mean=0.0271061655172, std=0.0644117728729
95.0% confidence resignation rate 0.0576625395852
Incorrect resignations in dataset with the suggested resign rate: 7.59%
Average game length without resignations: 412, after resignations: 299
Average game length reduction: 27.46%
89.66% games resigned
```

```
98.0% confidence resignation rate 0.0222428716957
Incorrect resignations in dataset with the suggested resign rate: 4.14%
Average game length without resignations: 412, after resignations: 335
Average game length reduction: 18.59%
78.62% games resigned
```

Enabling resignations even with a very low threshold would speed up the self-play games by about 20%. Number of incorrect resignations wouldn't be very high with the current network and it seems that the network can currently count with good enough accuracy.

If the clients would upload the minimum winners score during the self-play games it could be used to accurately and fast to set the correct resign rate for the each network on the server.

I uploaded the scripts I wrote to: https://github.com/Ttl/leela-zero/tree/resign_analysis/resign_analysis

**Comment by isty2e** [comment]
A nice experiment! The number of false positives would decrease with increase playouts, hopefully. 

By the way, how long did it take for this analysis? Could it be done for each training step without significant increase in computational cost?

**Comment by Ttl** [comment]
It takes few minutes to analyze the already played games. False positives currently seem to be caused by the network not realizing that it needs two eyes to live. Some games I looked at the winner's evaluation was very low because it didn't realize that one of the opponents groups is actually dead. After putting the group actually in atari the evaluation shoots up.

Note that during the analysis there are are no playouts and only the neural network output is used. It should be much better at evaluation when combined with the MCTS.

I'm not sure if resignation threshold can be determined from the validation games since one of the networks could be much stronger. In my opinion it should be determined from the self play games like in the AGZ paper. It wouldn't slow the self play games if the evaluations were recorded, but it needs adding a support for it in client and server codes.

**Comment by isty2e** [comment]
Maybe it can be done without such additional works: We have at least 10%, and practically more, of completed games. That means we'll have at least 50k fully completed games, and I suppose it will be sufficient for such an analysis. Of course we lose the playouts, but it can't be at least dangerous, I believe.

**Comment by jkiliani** [comment]
If autogtp automatically recorded the value head output after every move and sent this file to the server along with .sgf and training data (or, if the value head output was part of the training data), it would be a simple matter of statistically evaluating the minimum NN evaluation of the winner to determine a suitable resignation threshold. It would only add little to the data volume collected.

**Comment by isty2e** [comment]
Sure, it would be ideal if there is no problem with the storage.

**Comment by gcp** [comment]
Thanks for the analysis. It seems resigning at 3% would be reasonably safe, and if we disable resignation in 25% of the games that should give the net enough training data for the cases where it resigned wrongly. Both parameters seem robust enough to me that I'm wiling to enable this. 

That would give almost 20% speedup.

**Comment by Dorus** [comment]
Are we sure about this? In most self play games, the winner is correct even with premature passes, but i made some sneaky examples, for example with the [bulky five](https://senseis.xmp.net/?BulkyFive) and it doesn't recognize that yet. Is in the analyses above also checked if the correct winner was selected, even if a stronger bot got a chance to play it out? 

You don't even need the bulky five, i noticed that if a group is large enough and has a large eye with a opponent stone in it, even if it's only a 3 point eye, the value network doesn't think it can kill it yet. For now it seem to depend on the other player to reduce the liberties of this eye first before it captures. In most self play games this does happen eventually, but as the value network gets smarter, it might stop filling its own inside liberties and then incorrect resignations might happen.

(i'm just speculating a bit here, 25% of the games might be enough to teach it about these concepts anyway)

**Comment by betterworld** [comment]
It might be a good idea to embed a comment into the sgf for the 25% non-resignable games whenever the winrate drops below 3%. So if we aren't sure that it's safe, we can later grep for those comments and (hopefully) find out that most of the games were really lost.

**Comment by Dorus** [comment]
Another board: Here winrate does drop below 3% because white has a larger margin on the dead group.

http://eidogo.com/#CH7ruGF2

(;FF[4]CA[UTF-8]AP[GoGui:1.4.9]
KM[7.5]DT[2017-11-26]
AB[as][ar][aq][ap][ao][an][am][al][ak][aj][ai][ah][ag][af][ae][ad][ac][ab][aa][bs][br][bq][bp][bo][bn][bm][bl][bk][bj][bi][bh][bg][bf][be][bd][bc][bb][ba][cs][cr][cq][cp][co][cn][cm][cl][ck][cj][ci][cg][cf][ce][cd][cb][ca][ds][dr][dn][dm][dk][dj][di][dh][dg][df][de][dd][dc][db][da][es][er][eq][ep][en][em][el][ek][ej][ei][eh][eg][ef][ee][ed][ec][eb][ea][fs][fr][fq][fp][fn][fm][fl][fk][fj][fh][fg][fe][fd][fc][fb][fa][gs][gr][gq][gp][gn][gm][gl][gk][gj][gi][gh][gg][gf][hs][hr][hq][hp][hm][hl][hk][hj][hi][hh][hg][hf][is][ir][iq][ip][io][il][ik][ij][ii][ih][ig][js][jr][jq][jp][jo][jn][jl][jk][jj][ji][jh][kj][lj][mj][nj][oj][og][pj][qj][rj][sj]
AW[dq][dp][do][eo][fo][go][ge][gd][gc][gb][ga][ho][hn][he][hd][hc][hb][ha][in][im][if][ie][id][ic][ib][ia][jm][jg][jf][je][jd][jc][jb][ja][ks][kr][kq][kp][ko][kn][km][kl][kk][ki][kh][kg][kf][ke][kd][kc][kb][ka][ls][lr][lq][lp][lo][ln][lm][ll][lk][li][lh][lg][lf][le][ld][lc][lb][la][ms][mr][mo][mm][mk][mi][mh][mg][mf][me][md][mc][mb][ma][ns][nr][np][no][nn][nl][nk][ni][nf][ne][nd][nc][nb][na][os][or][oq][oo][on][om][ok][oi][of][oe][od][oc][ob][oa][ps][pr][pp][pn][pl][pk][pi][ph][pf][pe][pd][pc][pb][pa][qs][qr][qq][qo][qn][qm][qk][qi][qh][qg][qf][qe][qd][qc][qb][qa][rs][rr][rp][rn][rl][rk][ri][rh][rg][rf][re][rd][rc][rb][ra][ss][sr][sq][sp][so][sn][sm][sl][sk][si][sh][sg][sf][se][sd][sc][sb][sa]
PL[B])

**Comment by evanroberts85** [comment]
If shapes like the dead 5 are outside the range of the 1000 simulations, then the neural network is probably not ready to absorb that level of information anyway, so I do not think any harm is done by an early resignation. Leelaz is still learning the basics and does not understand liberties yet.

**Comment by Dorus** [comment]
I tested it with a 3 stone cut and it gave the same result.

The heatmap is simply unable to detect the stones are dead. Even if i reduce it to just a 1 point eye. It seems only the MCTS search is able to declare the stones dead by taking them. It seems it just assumes large groups are alive, if i try the same with a small group, it does consider it dead. Even to a point where it thinks it's acceptable to pass without cleaning up said stones.

**Comment by evanroberts85** [comment]
Leela must be taking the proverb “Big Dragons Never Die” literally 😎

**Comment by isty2e** [comment]
Do such counterexamples matter if the majority of the games are predicted correctly? We don't have to care about such specific cases nor force LZ to do something specific which a more skilled player would've done, I believe. Such skill issues can be ultimately handled by the RL process.

**Comment by Dorus** [comment]
We're not talking about inputing specific cases, rather, how likely adding resignation will make it  harder for the network to learn certain important concepts. Making games 20% shorter without losing quality is great. Making them 20% shorter by tossing 75% important date is terrible.

That's why I asks for an analysis of the current game with a stronger bot to see how often it would surrender correctly. If we just depend on the actual game result we will never spot bad results from early passes and surrenders.

**Comment by isty2e** [comment]
Again, I don't see any reason to do that. The "certain important concepts" you are talking about is something in human perspective, whereas the NN can learn other things you might not consider essential but still gain some skills. And I believe that we are to predict the results by the current nets, not by any stronger bots. That's how the reinforcement learning works, I believe.

**Comment by gcp** [comment]
The only concern is when the resigning would change the results between the current players, not what is theoretically optimal. If the bot resigns a game that a 9d bot could win, that doesn't matter as long as it would have lost against itself anyway, so nothing was changed by the resignation.

**Comment by killerducky** [comment]
The current code also requires the first child to have been visited more than 100 times. With only 1000 playouts a majority of the positions don't reach this limit. Maybe the 100 visits should be lowered because that was probably tuned for old-style playouts. 

Also Ttl's analysis was done on just a single pass of the NN, and the current resign code is based on the results of the 1000 playout UCT search. I have modified my local leelaz to save the NN's winrate, the UCT winrate, and the number of visits to the first child. By tomorrow I will have enough results to do more analysis.

@Ttl tomorrow I can send you the data I collect if you want to analyze it. Also I'll work on getting my code changes into a branch on github.

**Comment by tsuchiLo** [comment]
the shortest game.

Best network hash: 92c658d7325fe38f0c8adbbb1444ed17afd891b9f208003c272547a7bcb87909
Required client version: 2 (OK)
Already downloaded network.
Engine has started.
Infinite thinking time set.
1 (P5) 2 (pass) 3 (pass) Game has ended.
Score: B+353.5
Winner: black
Writing 3800028acf3d426ab112d832f23c41da.sgf
Dumping 3800028acf3d426ab112d832f23c41da.txt
Stopping engine.

**Comment by isty2e** [comment]
That is not the result of resignation, rather the consequence of double pass. I don't see why you are showing the short game here.

**Comment by gcp** [comment]
>Maybe the 100 visits should be lowered because that was probably tuned for old-style playouts.

Correct.

**Comment by killerducky** [comment]
I have some results but first a caveat: I used root.get_eval() to collect winrates, but I noticed I should use root.get_first_child()->get_eval() to match what the resign code does. I'll rerun this tonight.

So far using this close but not correct method, it looks like the threshold should be set higher when you use 1000 playouts. Presumably the net winrates are noisier requiring a lower threshold then the more stable uct winrates. 

"uct resigns" is based on root.get_eval(). "net resigns" is based on result.second where result =  Network::get_scored_moves(&state, Network::Ensemble::DIRECT, 0).

```
Dataset size 183 games
Resign rate: 0.50
Incorrect uct resigns = 182/183 (99.45)
Incorrect net resigns = 183/183 (100.00)

Resign rate: 0.20
Incorrect uct resigns = 18/183 (9.84)
Incorrect net resigns = 29/183 (15.85)

Resign rate: 0.15
Incorrect uct resigns = 11/183 (6.01)
Incorrect net resigns = 23/183 (12.57)

Resign rate: 0.10
Incorrect uct resigns = 6/183 (3.28)
Incorrect net resigns = 18/183 (9.84)

Resign rate: 0.05
Incorrect uct resigns = 2/183 (1.09)
Incorrect net resigns = 9/183 (4.92)
```

**Comment by sethtroisi** [comment]
@killerducky If you need extra games, now or in the future, I'm willing to run most arbitrary code (I can generate ~20 games / hour) and mail you results.

**Comment by killerducky** [comment]
I ran just a few games with root.get_first_child()->get_eval(), something is wrong, still analyzing...

**Comment by sethtroisi** [comment]
@killerducky Here's 60 games, hope it helps out: https://drive.google.com/open?id=1pFVt5pDce7sHes6kwS3enMz3ps4pnRfi

**Comment by killerducky** [comment]
New results with root.get_first_child()->get_eval(), and I also added code to calculate average game length. Note the 0.50 resign rate is in there as a sanity check, the games average 2 moves.

I didn't analyze number visits for best child, although the data is there. It might be simplest to change it to number visits for the root node > 500. If we keep it as best child visits, it does go over 100 often, but lowering it would catch more. Just glancing at it seems like 50 or 30 would catch most.

Also I just checked the paper, probably none of these details matter but:
> AlphaGo Zero resigns if its root value and best child value are lower than a threshold value vresign.

```
Resign rate: 0.50
Incorrect uct resigns = 48/182 (26.37%)
Average game length = 409. Average game length with resigns = 2 (99.42% reduction)

Resign rate: 0.20
Incorrect uct resigns = 11/182 (6.04%)
Average game length = 409. Average game length with resigns = 290 (28.96% reduction)

Resign rate: 0.15
Incorrect uct resigns = 8/182 (4.40%)
Average game length = 409. Average game length with resigns = 307 (24.90% reduction)

Resign rate: 0.10
Incorrect uct resigns = 5/182 (2.75%)
Average game length = 409. Average game length with resigns = 326 (20.36% reduction)

Resign rate: 0.05
Incorrect uct resigns = 2/182 (1.10%)
Average game length = 409. Average game length with resigns = 348 (14.81% reduction)
```

**Comment by isty2e** [comment]
I wonder when it will be adopted in the live version... Maybe it will be handled on the server, not client, like [this](https://github.com/gcp/leela-zero/issues/153#issuecomment-348786075)?
